### ISSUE TITLE
Race conditions in checker when I use parapoly + explicit overloading

### ISSUE THREAD

**Issue by karl-zylinski** [odin-lang/Odin#2968]
**Edit: Jakub has created a repro for this, see his comment on this issue!**

I do not have a good repro for this outside of my big code base, but this is something me and others like (@jakubtomsu and @laytan) has seen, which is why I'm opening an issue so we can track this and add context over time.

I randomly get compiler crashes like these whenever I use procs that have parapoly, combined with putting those procs into an explicitly overloaded proc:
`C:\Programs\Odin\src\llvm_backend_type.cpp(12): Panic: NOT FOUND lb_type_info_index ^EntityVariant @ index 267`

My overloaded proc might look something like (snipped from my serializer):

```
ser :: proc {
	ser_int,
	ser_slice,
	ser_dynamic_array,
	ser_fixed_array,
	ser_bool,
	ser_float,
	ser_enum,
	ser_string,

	// more complex types
	ser_entity,
	ser_entity_variant,
}
```

and for example an object might contain a slice, which runs `ser_slice`, which is a parapoly proc, which will then run `ser` again to figure out what to do with the generic slice elements. It is in those cases that the compiler randomly blows up.

The compiler crashes seem like race conditions.

These days I always compile with `-no-threaded-checker`, which makes these problems go away at the cost of compile time.

I do not currently have any kind of minimal repro, I have tried to make one but it never happens outside of my big program.

I remember trying @jakubtomsu's LBP serializer (which I based my serializer upon) https://github.com/jakubtomsu/odin-lbp-serialization/tree/main, and just adding a few overloads to `serialize` in there could make this start happen randomly.

@jakubtomsu, @laytan you saw similar issues, do you have context to add?

* Operating System & Odin Version:
        Windows, Git master version
* Please paste `odin report` output:
        Odin: dev-2023-11:0340a9d9
        OS:   Windows 10 Professional (version: 22H2), build 19045.3693
        CPU:  Intel(R) Core(TM) i7-6950X CPU @ 3.00GHz
        RAM:  32674 MiB

**Comment by jakubtomsu** [comment]
Ok I think I have come up with a good repro. On my PC it has 40-50% failure rate, but that probably depends on the core count.

This code is based on my [LBP serializer](https://github.com/jakubtomsu/odin-lbp-serialization), but heavily simplified. I removed everything that didn't have an effect on the error.

```odin
package game
// Explanation of the LBP serialization method:
// https://handmade.network/p/29/swedish-cubes-for-unity/blog/p/2723-how_media_molecule_does_serialization

import "core:fmt"
import "core:intrinsics"
import "core:mem"
import "core:slice"

main :: proc() {
    s: Serializer
    foo: Foo
    serialize(&s, &foo)
}

serialize :: proc {
    serialize_basic,
    serialize_bit_set,
    serialize_array,
    serialize_slice,
    serialize_string,
    serialize_dynamic_array,

    // Custom
    serializer_foo,
    serialize_bar,
}

Foo :: struct {
    bars: []Bar,
}

Bar :: struct {}

serializer_foo :: proc(s: ^Serializer, foo: ^Foo) -> bool {
    return serialize(s, &foo.bars)
}

serialize_bar :: proc(s: ^Serializer, bar: ^Bar) -> bool {
    return true
}

Serializer :: struct {}

serializer_debug_scope :: proc(s: ^Serializer, name: string) {}

serialize_basic :: proc(
    s: ^Serializer,
    data: ^$T,
) -> bool where intrinsics.type_is_float(T) ||
    intrinsics.type_is_integer(T) ||
    intrinsics.type_is_enum(T) ||
    intrinsics.type_is_boolean(T) {
    return true
}

serialize_slice :: proc(s: ^Serializer, data: ^$T/[]$E) -> bool {
    serializer_debug_scope(s, fmt.tprint(typeid_of(T))) // Removing this line stops the issue for some reason
    return true
}

serialize_bit_set :: proc(s: ^Serializer, data: ^$T/bit_set[$E]) -> bool {return true}
serialize_array :: proc(s: ^Serializer, data: ^$T/[$S]$E) -> bool {return true}
serialize_string :: proc(s: ^Serializer, data: ^string) -> bool {return true}
serialize_dynamic_array :: proc(s: ^Serializer, data: ^$T/[dynamic]$E) -> bool {return true}
serialize_map :: proc(s: ^Serializer, data: ^$T/map[$K]$V) -> bool {return true}
```
The error is this:
```
D:\a\Odin\Odin\src\llvm_backend_type.cpp(12): Panic: NOT FOUND lb_type_info_index []Bar @ index 184
```
I also made a runner in python to quickly test changes:
```py
import os
while True:
    os.system("odin build test.odin -file")
```

**Comment by karl-zylinski** [comment]
Amazing, thanks for the repro!

**Comment by JesseRMeyer** [comment]
There's a lot of hazards currently in the compiler's threaded path.  Mini-repos are helpful because whenever someone takes an actionable look at these problems, we can test the fixes against all of them at once.  Below I did the basic thing of just running the compiler with -fsanitzer=thread and boy was the output noisy.  So hopefully there's low hanging fruit where a few surgical fixes resolve these issues together.

https://github.com/odin-lang/Odin/issues/2949#issuecomment-1817959490

**Comment by karl-zylinski** [comment]
Fixed by @gingerBill  in commit d8bb93accc3793b82609f2b14881272f3e19af27
